lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU

```diff
@@ -35,123 +39,111 @@ VAR
     '}
-    fCalibRelRatio: LREAL;
+    uAccelerator: UINT;
 
     {attribute 'pytmc' := '
-        pv: CALIB:MJ_RATIO
-        io: io
+        pv: NCBeamRateFromEpics
+        link: TPG:SYS0:1:DST04:RATE
     '}
-    fCalibMJRatio: LREAL;
+    uNCFrequency: UINT;
 
-    // Linked to beam rate
     {attribute 'pytmc' := '
-        pv: BeamRateFromEpics
+        pv: SCBeamRateFromEpics
         link: TPG:SYS0:1:DST04:RATE
     '}
-    fFrequency: UINT;
+    uSCFrequency: UINT;
+
+    uFrequency: UINT;
 
     // Pulse-by-pulse energy
     {attribute 'pytmc' := '
-        pv: ResponMJ
+        pv: PulseEnergy
         io: i
         field: EGU mJ
     '}
-    fResponMJ: LREAL;
+    fCalibMJ: LREAL;
 
     // Pulse-by-pulse energy buffer
     {attribute 'pytmc' := '
-        pv: ResponMJ_BUFFER
+        pv: PulseEnergy_BUFFER
         io: i
         field: EGU mJ
     '}
-    fResponMJBuffer: ARRAY[1..1000] OF LREAL;
+    fCalibMJBuffer: ARRAY[1..1000] OF LREAL;
 
     // Wattage
     {attribute 'pytmc' := '
-        pv: ResponWattage
+        pv: PulseWattage
         io: i
         field: EGU W
     '}
-    fResponWattage: LREAL;
+    fPulseWattage: LREAL;
 
     // Background voltage
     {attribute 'pytmc' := '
-        pv: Background
+        pv: BackgroundVoltage
         io: io
         field: EGU V
     '}
-    fBackground: LREAL;
+    fBackgroundVoltage: LREAL;
+
+    fBackgroundVoltageAcc: LREAL;
+    fBackgroundVoltageSum: LREAL;
+    fBackgroundVoltageStale: LREAL;
+    tonBackgroundAutoCollecting: TON;
+    fBackgroundVoltageBuffer: ARRAY[1..100] OF LREAL;
+    udBackgroundVoltageBufferIndex: UDINT := 1;
+    i: UDINT;
+
+    {attribute 'pytmc' := '
+        pv: BackgroundMode
+        io: io
+    '}
+    BACKGROUND_MODE: (Manual, Passive) := Passive;
 
     // Boolean to trigger collection of background voltages
     {attribute 'pytmc' := '
         pv: BackgroundCollect
         io: i
     '}
-    fBackgroundCollect: BOOL;
+    bBackgroundCollect: BOOL;
 
     {attribute 'pytmc' := '
         pv: BackgroundCollectionTime
         io: io
         field: EGU s
     '}
     // Time in seconds to collect background voltages for
-    fBackgroundCollectionTime: UINT;
-    // Number of cycles to collect background voltages for
-    fBackgroundCollectionCycles: UDINT;
-    // Number of cycles that background voltages have already been collected
-    fBackgroundCollectionCyclesDone: UDINT;
-    // Sum of background voltages, should only be used internally
-    fBackgroundCollectionSum: LREAL;
+    uBackgroundCollectionTime: UINT := 60;
+
+    tofBackgroundCollecting: TOF;
+    udBackgroundManualCount: UDINT;
 
     fbGetPMVoltage: FB_AnalogInput;
+
     fbVoltageBuffer: FB_LREALBuffer;
-    fbCalibBaseBuffer: FB_LREALBuffer;
     fbCalibMJBuffer: FB_LREALBuffer;
-    fbResponMJBuffer: FB_LREALBuffer;
+
+    FFO: FB_FastFault :=(
+        i_Desc := 'Fault occurs when the temprature trip point is reached',
```

Comment:
```suggestion
        i_Desc := 'Fault occurs when the temperature trip point is reached',
```

Reply:
Aside from the typo, maybe this description should be more clear that this is about the per-pulse MJ reading being too high for the powermeter? I'm not 100% sure how to phrase this, and it technically is a temperature, but I suspect that calling it a temperature will confuse people who are looking at the other thermocouples on that device that give us readouts in Celcius.

Reply:
How about "Fault occurs when the per-pulse energy reading is high enough to damage the power meter?" Also, do you know of a better typecode for this? I meant to use 16#400, not 16#f400, which is the optics code based on https://confluence.slac.stanford.edu/display/L2SI/PMPS+Type+Codes.

Reply:
I think just that's a good description.
For the typecode: I think you should pick a new number close to a similar fault and add it to the confluence table. I think this is functionally distinct enough from the other codes that it deserves a new number.

Reply:
Added code 16#500 - Pulse energy measured by PPM is high enough to damage power meter.